x/internal/outline/outline_packet_proxy.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package outline
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/Jigsaw-Code/outline-sdk/network"
+	"github.com/Jigsaw-Code/outline-sdk/network/dnstruncate"
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/x/connectivity"
+)
+
+type OutlinePacketProxy interface {
+	network.PacketProxy
+
+	TestConnectivityAndRefresh(resolver, domain string) error
+}
+
+type outlinePacketProxy struct {
+	network.DelegatePacketProxy
+
+	remotePktListener transport.PacketListener // this will be used in connectivity test
+	remote, fallback  network.PacketProxy
+}
+
+var _ OutlinePacketProxy = (*outlinePacketProxy)(nil)
+
+func NewOutlinePacketProxy(config *SessionConfig) (opp OutlinePacketProxy, err error) {
+	proxy := outlinePacketProxy{}
+
+	proxy.fallback, err = dnstruncate.NewPacketProxy()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create DNS truncate proxy: %w", err)
+	}
+
+	// Create Shadowsocks UDP PacketProxy
+	proxy.remotePktListener, err = NewOutlinePacketListener(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create UDP listener: %w", err)
+	}
+
+	proxy.remote, err = network.NewPacketProxyFromPacketListener(proxy.remotePktListener)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create UDP proxy: %w", err)
+	}
+
+	// Create DelegatePacketProxy
+	proxy.DelegatePacketProxy, err = network.NewDelegatePacketProxy(proxy.fallback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create delegate UDP proxy: %w", err)
+	}
+
+	return &proxy, nil
+}
+
+func (proxy *outlinePacketProxy) TestConnectivityAndRefresh(resolver, domain string) error {
```

Comment:
This leaks implementation details of the connectivity test to the user of the outlinePacketProxy. If we no longer use DNS resolution, the user will need to change. That's the kind of dependency that makes it hard to change things.

Can you provide just a Refresh() with no arguments?

Perhaps you pass the connectivity test as a function in the constructor?
That way you also won't need to hold the `remotePktListener`, which is a bit awkward.

Reply:
Instead of passing a connectivity test function, I'm thinking of whether an `TransportOptions` object would be more readable?

```
type PacketTransportOptions struct {
  PacketListener transport.PacketListener
  TestConnectivity func() error
}

func NewOutlinePacketProxy(options PacketTransportOptions) (network.PacketProxy, error) {
  // ...
}
```

Reply:
Why have an  "options" structure if you can have the actual object directly?
Also, if we are talking about the app-specific Outline Device, we can probably just hardcode the connectivity test, since our app doesn't customize it..